Sources/TestingInternals/Discovery.cpp

```diff
@@ -248,105 +322,108 @@ static SWTMachHeaderList getMachHeaders(void) {
     return true;
-  }, const_cast<SectionEnumerator *>(&body));
+  }, const_cast<ImageEnumerator *>(&body));
 }
 #endif
 
 #pragma mark -
 
 void swt_enumerateTypes(void *context, SWTTypeEnumerator body, SWTTypeNameFilter nameFilter) {
-  enumerateTypeMetadataSections([=] (const void *section, size_t size) {
-    auto records = reinterpret_cast<const SWTTypeMetadataRecord *>(section);
-    size_t recordCount = size / sizeof(SWTTypeMetadataRecord);
-
-    for (size_t i = 0; i < recordCount; i++) {
-      const auto& record = records[i];
-
-      auto contextDescriptor = record.getContextDescriptor();
-      if (!contextDescriptor) {
-        // This type metadata record is invalid (or we don't understand how to
-        // get its context descriptor), so skip it.
-        continue;
-      } else if (contextDescriptor->isGeneric()) {
-        // Generic types cannot be fully instantiated without generic
-        // parameters, which is not something we can know abstractly.
-        continue;
-      }
+  enumerateImages([=] (const SWTMetadataSections& sections) {
+    // If this image has a tests section, then it should be used instead of the
+    // image's type metadata section.
+    if (sections.version >= SWT_METADATA_SECTION_MINIMUM_VERSION_WITH_TESTS
+        && sections.swift5_tests.start != 0
+        && sections.swift5_tests.length > 0) {
+      return; // continue
+    }
 
-      // If the caller supplied a name filtering function, check that the type's
-      // name passes. This will be more expensive than the checks above, but
-      // should be cheaper than realizing the metadata.
-      if (nameFilter) {
-        const char *typeName = contextDescriptor->getName();
-        bool nameOK = typeName && (* nameFilter)(typeName, context);
-        if (!nameOK) {
+    if (auto records = reinterpret_cast<const SWTTypeMetadataRecord *>(sections.swift5_type_metadata.start)) {
+      size_t recordCount = sections.swift5_type_metadata.length / sizeof(SWTTypeMetadataRecord);
+      for (size_t i = 0; i < recordCount; i++) {
+        const auto& record = records[i];
+
+        auto contextDescriptor = record.getContextDescriptor();
+        if (!contextDescriptor) {
+          // This type metadata record is invalid (or we don't understand how to
+          // get its context descriptor), so skip it.
+          continue;
+        } else if (contextDescriptor->isGeneric()) {
+          // Generic types cannot be fully instantiated without generic
+          // parameters, which is not something we can know abstractly.
           continue;
         }
+
+        // If the caller supplied a name filtering function, check that the type's
+        // name passes. This will be more expensive than the checks above, but
+        // should be cheaper than realizing the metadata.
+        if (nameFilter) {
+          const char *typeName = contextDescriptor->getName();
+          bool nameOK = typeName && (* nameFilter)(typeName, context);
```

Comment:
Nit: extra space.

```suggestion
          bool nameOK = typeName && (*nameFilter)(typeName, context);
```

Reply:
House style (not that it comes up much.) Well… my style, anyway, because I am old and `fp()` was, historically, not a valid way to invoke a function pointer in C… anyway, I put an onion on my belt, which was the style at the time…